runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -294,16 +294,43 @@ protected:
       }
 
       // User state provided...
-
-      // FIXME handle case where pointer is a device pointer
+      // Check if the pointer is a device pointer
+      cudaPointerAttributes attributes;
+      HANDLE_CUDA_ERROR(cudaPointerGetAttributes(&attributes, state));
 
       // First allocation, so just set the user provided data here
       ScopedTraceWithContext(
           "CuStateVecCircuitSimulator::addQubitsToState cudaMemcpy",
           stateDimension * sizeof(CudaDataType));
-      HANDLE_CUDA_ERROR(cudaMemcpy(deviceStateVector, state,
+
+      if (attributes.type == cudaMemoryTypeDevice) {
+        int currentDevice;
+        HANDLE_CUDA_ERROR(cudaGetDevice(&currentDevice));
+
+        if (attributes.device != currentDevice) {
```

Comment:
This was discussed offline, but just to capture it here, we think it would be good to add a multi-GPU test somewhere to exercise this new code in this if branch; otherwise, the code in here would never have been tested, and there is some question about whether or not the calls to `cudaSetDevice` would work as intended here.

Reply:
Another alternative would be to throw an "unsupported" error message here, but I think we should still add a test for the "same GPU" functionality as well.